Transaction validity and to-plain conversion for PoW transactions

sisou opened the pull request:
Ensure PoW transactions can be properly converted to their plain representation, which is primarily done during address transaction history fetching in the web-client. Before this change, PoW transactions could not be returned and the process panicked, due to different serializations of contract creation data, HTLC proofs and signature proofs.

I have tried to document all changes and differences in code comments, let me know if you require more info, here or in the code.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

nibhar committed:
Prepend a 0 to PoW transaction Proofs. They lack the `webauthn_fields`

nibhar committed:
Make sure not to ask for block inclusion proof of genesis block

sisou committed:
Enable deserializing PoW contract creation data

PoW contract creation data was different than in PoS, leading to failing deserializations when converting a transaction to its plain format in the web-client.

sisou committed:
Pass genesis parameters to contract creation data conversion when needed

sisou committed:
Skip attempting to parse PoW HTLC claiming proofs for now

sisou committed:
Use a custom struct to properly parse and convert PoW signature proofs

sisou committed:
WIP: Parse and convert PoW HTLC claiming proofs

jsdanielh committed:
Implement deserialization of PoW Outgoing HTLC transaction proofs

sisou committed:
Correct PoW signature proof deserialization

sisou committed:
Ensure PoW transactions `.verify()`

Added special handling for PoW network IDs and added tests for basic txs, vesting and HTLC creation txs and HTLC claiming txs.

sisou committed:
Update network ID in tests and doctests to MainAlbatross

Where it makes sense.

sisou committed:
Implement custom deserializer for PoW multisig signature proofs

The problem was the merkle path, whose serialization changed in PoS.

jsdanielh commented on pow-migration/Cargo.toml:
```diff
@@ -72,6 +72,7 @@ build-data = "0.2"
 serde_json = "1.0"
+nimiq-web-client = { workspace = true, default-features = false, features = ["primitives"] }
 
 nimiq-test-log = { workspace = true }
```
```suggestion

nimiq-test-log = { workspace = true }
nimiq-web-client = { workspace = true, default-features = false, features = ["primitives"] }
```

jsdanielh commented on primitives/transaction/src/account/htlc_contract.rs:
```diff
@@ -225,6 +243,49 @@ impl CreationTransactionData {
 }
 
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct PoWCreationTransactionData {
```
Could you add some `rustdoc`?

jsdanielh commented on primitives/transaction/src/account/htlc_contract.rs:
```diff
@@ -225,6 +243,49 @@ impl CreationTransactionData {
+        Ok(())
+    }
+
+    pub fn into_pos(self, genesis_number: u32, genesis_timestamp: u64) -> CreationTransactionData {
```
```suggestion
    pub fn into_pos(self, genesis_block_number: u32, genesis_timestamp: u64) -> CreationTransactionData {
```

jsdanielh commented on primitives/transaction/src/account/htlc_contract.rs:
```diff
@@ -340,6 +401,129 @@ impl OutgoingHTLCTransactionProof {
 }
 
+#[derive(Clone, Debug)]
+pub struct PoWRegularTransfer {
```
Could you add some `rustdoc`?

jsdanielh commented on primitives/transaction/src/account/vesting_contract.rs:
```diff
@@ -198,3 +206,120 @@ impl CreationTransactionData {
+            _ => return Err(TransactionError::InvalidData),
+        })
+    }
+    pub fn parse(transaction: &Transaction) -> Result<Self, TransactionError> {
```
```suggestion
    }

    pub fn parse(transaction: &Transaction) -> Result<Self, TransactionError> {
```

jsdanielh commented on primitives/transaction/src/account/vesting_contract.rs:
```diff
@@ -198,3 +206,120 @@ impl CreationTransactionData {
+        PoWCreationTransactionData::parse_data(&transaction.recipient_data, transaction.value)
+    }
+
+    pub fn into_pos(self, genesis_number: u32, genesis_timestamp: u64) -> CreationTransactionData {
```
```suggestion
    pub fn into_pos(self, genesis_block_number: u32, genesis_timestamp: u64) -> CreationTransactionData {
```

jsdanielh commented on primitives/transaction/src/signature_proof.rs:
```diff
@@ -350,6 +350,42 @@ impl WebauthnExtraFields {
 }
 
+#[derive(Clone, Debug, Deserialize)]
+pub struct PoWSignatureProof {
```
Could you add some `rustdoc`?

sisou committed:
Add documentation to PoW structs

Also
- refactor `genesis_number` => `genesis_block_number`
- clean up example page changes

sisou committed:
DRY split between PoW and PoS contract creation data parsing

sisou commented:
Ready for review.

jsdanielh commented on pow-migration/Cargo.toml:
```diff
@@ -74,6 +74,7 @@ build-data = "0.2"
 serde_json = "1.0"
 
 nimiq-test-log = { workspace = true }
+nimiq-web-client = { workspace = true, default-features = false, features = ["primitives"] }
```
The  `default-features = false` is inherited from the workspace
```suggestion
nimiq-web-client = { workspace = true, features = ["primitives"] }
```
